doc/smidump.1
.\"
.\" $Id$
.\"
.TH smidump 1  "May 5, 1999" "IBR" "SMI Tools"
.SH NAME
smidump \- dump and convert SMIv1/v2/ng modules
.SH SYNOPSIS
.B smidump
[
.B "-sSvV"
] [
.BI "-l " level
] [
.BI "-d " level
] [
.BI "-c " config
] [
.BI "-L " location
] [
.BI "-D " format
]
.I "module"
.SH DESCRIPTION
The \fBsmidump\fP program is used to dump the contents of a single MIB
module to stdout in a selectable format. This format may be a simple
list of objects or types but also a format fully compliant to SMIv2
or SMIng. This allows to convert modules from SMIv2 to SMIng and from
SMIng to SMIv2.
.PP
\fBSmidump\fP first reads the global libsmi configuration file
\fB/usr/local/etc/smi.conf\fP, as any libsmi application does. Then
the command line options are interpreted and finally, the
\fB/usr/local/etc/smidump.conf\fP configuration file is read, if present,
before \fBsmidump\fP starts its main operation.
.SH OPTIONS
.TP
.B "-s"
Report some per module statistics to stderr.
.TP
.B "-S"
Don't report statistics (default).
.TP
.B "-v"
On errors, do print the current line of input to stderr after the
error message and a marker pointing to the current token in the input
stream. Note that the corresponding point of error usually lies before
this marker.
.TP
.B "-V"
Don't print error lines (default).
.TP
.BI "-d " level
Print libsmi debugging messages at the given \fIlevel\fP (default: 0).
.TP
.BI "-l " level
Report errors and warnings up to the given \fIlevel\fP to stderr.
A higher level shows more warnings. The actual levels are:
.sp
.nf
 0   internal error, no recovery possible,
 1   major SMI error, recovery somehow possible but
     may lead to severe problems,
 2   SMI error, probably tolerated by some
     implementations,
 3   SMI error, tolerated by many implementations,
 4   something which is recommended to be changed,
 5   something that might be ok, but might not be
     recommended under some circumstances.
.fi
.sp
Higher levels are actually not used and lead to the same effects as
level 5 does. Note that errors up to level 3 are errors violating the
specification and must be fixed by the responsible author. Level 3 is
the default level.
.TP
.BI "-c " config
Read the configuration from \fIfile\fP. See smi_config(3) section
CONFIGURATION FILES for its format.
.TP
.BI "-L " location
Add \fIlocation\fP to the list of locations to search for MIB modules.
See smi_config(3) section MODULE LOCATIONS for location details.
.TP
.I module
This is the module to be checked. Note that this module is searched in
the list of known locations. If the module shall be taken from an
SMIv1/v2 file in the current directory, a "\fB-L smi:.\fP" option
should be added to the command line.
.SH "EXAMPLE"
This example converts the SMIv2 module IF-MIB in the current directory
with the highest parser error verbosity to IF-MIB.sming in SMIng format.
.nf

  $ smidump -Lsmi:. -l9 -Dsming IF-MIB > IF-MIB.sming
.fi
.SH "FILES"
.nf
/usr/local/etc/smi.conf     general configuration file
/usr/local/etc/smidump.conf smidump configuration file
.fi
.SH "NOTES"
Libsmi and this tool are work in progress. Expect changes in near
future.
.SH "SEE ALSO"
The SMIng and
.BR libsmi (3)
project is documented at
.BR "http://www.ibr.cs.tu-bs.de/~strauss/sming/" "."
.PP
.BR smilint "(1)"
.SH "AUTHOR"
Frank Strauss <strauss@ibr.cs.tu-bs.de>
.br


ChangeLog
2000-11-24  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* libsmi-0.2.10.tar.gz: released version 0.2.10.

	* configure.in: Fixed broken 64 bit type configuration.

2000-11-20  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* changed filenames of file created by the corba and netsnmp
	dump drivers. Adapted test suite.

	* libsmi-0.2.9.tar.gz: released version 0.2.9.

2000-11-16  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* smi.conf-example: Added a configuration file example.

	* configure.in: Dir separator is now configurable and distinguished
	in util/smiIsPath().

	* win/*: Some changes for clean MSVC compilation. Yigal.

2000-11-14  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* win/config.h: Added macros for 64 bit types on MSVC systems.
	Added a necessary compiler flag to compile scanner-*.c on MSVC.
	Thanks to Yigal.

	* configure.in: Fixed: completely broken AC_TRY_RUN check for
	64 bit types. Somehow, I did it in a way that was syntactically
	correct. ;-)

	* acconfig.h: renamed MIN/MAX macros to suppress redefinitions
	on Solaris 2.6+.

2000-11-13  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* libsmi-0.2.8.tar.gz: released version 0.2.8.

	* test/Makefile.am: Added tests for formats: identifiers,
	metrics, xml, cm, python.
	
	* tools/dump-jax.c: Fixed dangling pointers. Test suite
	is happy again.

	* tools/dump-corba.c: The driver now creates idl and oid files
	instead of writing to stdout.
	
	* test/*: Reorganized the test suite. All scripts for
	smidump scripts are now identical. Added jax test (and found
	a bug that is not yet fixed).

2000-11-12  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* tools/dump-netsnmp.c: Renamed dump-ucdsnmp.c to dump-netsnmp.c.

	* test/Makefile.am: Commented out checks for the obsolete
	corba-{idl,oid} checks to make the test suite happy.

	* configure.in: The availability of `long long' type is now
	checked during configuration. The result now also affects printf
	formats.

	* configure.in: Path separator is now configurable (useful to
	use ":" instead of ";" as well in cygwin environments).

2000-11-09  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* tools/smidump.c: Juergen did a major rework of the internal
	smidump interface for the output drivers. Visible changes:
	corba-idl and corba-oid are integrated into a single format
	corba; -o option.

2000-11-08  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/iana/*: Updated IANA maintained modules.

	* lib/parser-smi.y: Suppressed an module conformance error
	message in case of SMIv1.

2000-11-08  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* lib/check.c,parser-smi.y: Bill Fenner fixed some duplicated
	error messages.

	* tools/dump-smi.c,dump-tree.c: Bill Fenner fixed some core
	dumps on erroneous MIB modules.

	* tools/dump-ucdsnmp.c: Juergen changed the UCD-SNMP driver
	so that it now creates three file (header, stubs & implementation)
	with a single smidump format option (netsnmp).

	* lib/parser-smi.y: Added checks for illegal subtyping; adjusted
	test suite, so that is runs successfully for all current tests.

	* lib/scanner-smi.l,scanner-sming.l: Bill Fenner contributed
	a patch to continue scanning after unexpected characters have
	been read.

2000-11-06  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* libsmi-0.2.7.tar.gz: released version 0.2.7.

	* lib/parser-smi.y: Fixed `0' ranges for some core types on Solaris.

	* lib/smi_config.3.in: Updated the library man pages.

	* tools/smidump.c: addModule() now inserts NULL module at the
	head of the list as expected by the dump drivers.

	* lib/check.c: Added checks for OID redifinitions and recursive
	definitions, based on a contribution from Bill Fenner.

	* lib/smi.h.in: Added a SmiModule.conformance attribute that
	allows applications to determine, `how syntactically correct'
	a module is.

2000-11-04  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/DISMAN-EVENT-MIB: added (RFC 2981).

	* mibs/ietf/DISMAN-EXPRESSION-MIB: added (RFC 2982). It needed
	a small fix.

2000-11-01  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* lib/smi.c: Reworked the smiPath initialization upon a hint
	from Bill Fenner. Note that the order of evaluation of the SMIPATH
	environment variable and configuration files has changed and that
	the syntax of `path' commands in configuration files has changed.
	See smi_config(3), section MODULE LOCATIONS for details.

2000-10-27  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* lib/check.c: Juergen fixed index checks for Bits/fixed-length
	octet strings.

2000-10-25  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* tools/smidump.c: Removed again `-X' option to eliminate parser
	options that might lead to different results.

2000-10-21  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* lib/error.c: Added a fourth arg to the error handler representing
	the error name. This has been wished and contributed by Bill Fenner.
	Note that this changed the API slightly.

2000-10-20  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* lib/check.c: Fixed index length checks for OCTET STRINGs,
	IpAddresses and OBJECT-IDENTIFIERs. Thanks to Bill Fenner.

	* lib/error.c: Fixed two minor typos. Thanks to Bill.

2000-10-19  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* tools/dump-python.c: a new dump format to generate Python MIB
	dictionary code. Contributed by Pat Knight.

	* tools/dump-xml.c: Fixed missing end tag of rows and tables
	in certain cases. Contributed by Pat Knight.

	* tools/smidump.c: Added option `-X' for `lax' parsing. Some
	people wish to accept things like underscores or upper case
	first letters in object type names. Note that it is usually
	a *bad* idea to use lax parsing!

2000-10-18  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/FRNETSERV-MIB: updated (RFC 2954).

	* mibs/ietf/FR-ATM-PVC-SERVICE-IWF-MIB: added (RFC 2955).

2000-10-17  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* tools/dump-ucdsnmp.c: Juergen fixed BITS handling.

2000-10-12  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/RTP-MIB: added (RFC 2959).


trunk/tools/smidump.c
/*
 * smidump.c --
 *
 *      Dump a MIB module conforming to a given format.
 *
 * Copyright (c) 1999 Frank Strauss, Technical University of Braunschweig.
 * Copyright (c) 1999 J. Schoenwaelder, Technical University of Braunschweig.
 *
 * See the file "COPYING" for information on usage and redistribution
 * of this file, and for a DISCLAIMER OF ALL WARRANTIES.
 *
 * @(#) $Id: smidump.c,v 1.42 2000/07/04 10:07:10 strauss Exp $
 */

#include <config.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif
#ifdef HAVE_GETOPT_H
#include <getopt.h>
#endif
#ifdef HAVE_WIN_H
#include "win.h"
#endif

#include "smi.h"
#include "smidump.h"



typedef struct {
    char *name;				/* Name of the output driver. */
    void (*func) (Module *);		/* Output generating function. */
    int smiflags;			/* Flags for the SMI parser. */
    int ignflags;			/* Output driver flags ignored. */
    char *descr;			/* Short description. */
} Driver;
 


static Driver driverTable[] = {
    { "SMIng",	   dumpSming,	0,			SMIDUMP_FLAG_UNITE,
      "SMIng" },
    { "SMIv2",	   dumpSmiV2,	0,			SMIDUMP_FLAG_UNITE,
      "SMIv2 (RFC 2578, RFC 2579, RFC 2580)"},
    { "SMIv1",	   dumpSmiV1,	0,			SMIDUMP_FLAG_UNITE,
      "SMIv1 (RFC 1155, RFC 1212, RFC 1215)"},
    { "imports",   dumpImports,	SMI_FLAG_NODESCR,	SMIDUMP_FLAG_UNITE,
      "recursive list of all imports" },
    { "types",	   dumpTypes,	SMI_FLAG_NODESCR,	0,
      "recursive list of all derived types" },
    { "tree",	   dumpTree,	SMI_FLAG_NODESCR,	0,
      "structure of the OID tree" },
    { "identifiers",dumpIdentifiers,SMI_FLAG_NODESCR,	0,
      "list of all identifiers" },
    { "metrics",   dumpMetrics,	0,			0,
      "metrics characterizing MIB modules" },
    { "mosy",	   dumpMosy,	SMI_FLAG_NODESCR,	SMIDUMP_FLAG_UNITE,
      "intermediate format generated by the mosy compiler" },
    { "xml",       dumpXml,	0,			SMIDUMP_FLAG_UNITE,
      "intermediate SMI XML exchange format" },
    { "corba-idl", dumpCorbaIdl, 0,			SMIDUMP_FLAG_UNITE,
      "corba IDL interface definitions (JIDM)" },
    { "corba-oid", dumpCorbaOid, SMI_FLAG_NODESCR,	SMIDUMP_FLAG_UNITE,
      "corba OID definitions (JIDM)" },
    { "cm-dia",	   dumpCMDia,	SMI_FLAG_NODESCR,	0,
      "reverse engineered conceptual model, dia format" },
    { "cm-xplain", dumpCMXplain,SMI_FLAG_NODESCR,	0,
      "reverse engineered conceptual model, ASCII format"},
    { "ucd-h",     dumpUcdH,	SMI_FLAG_NODESCR,	SMIDUMP_FLAG_UNITE,
      "UCD SNMP mib module C header" },
    { "ucd-c",	   dumpUcdC,	SMI_FLAG_NODESCR,	SMIDUMP_FLAG_UNITE,
      "UCD SNMP mib module C code" },
    { "jax",       dumpJax,	SMI_FLAG_NODESCR,	SMIDUMP_FLAG_UNITE,
      "Java AgentX sub-agent classes in separate files" },
#if 0
    { "sql",       dumpSql,	SMI_FLAG_NODESCR,	SMIDUMP_FLAG_UNITE,
      "SQL data model definitions" },
#endif
#if 0
    { "fig-tree",  dumpFigTree,	SMI_FLAG_NODESCR,	SMIDUMP_FLAG_UNITE,
      "tree graphics in xfig fig format" },
    { "fig-uml",   dumpFigUml,	SMI_FLAG_NODESCR,	SMIDUMP_FLAG_UNITE,
      "UML graphics in xfig fig format" },
#endif
    { NULL, NULL, 0, 0, NULL }
};



static Module *moduleList = NULL;



void *xmalloc(size_t size)
{
    char *m = malloc(size);
    if (! m) {
	fprintf(stderr, "smidump: malloc failed - running out of memory\n");
	exit(1);
    }
    return m;
}



void *xrealloc(void *ptr, size_t size)
{
    char *m = realloc(ptr, size);
    if (! m) {
	fprintf(stderr, "smidump: realloc failed - running out of memory\n");
	exit(1);
    }
    return m;
}



char *xstrdup(const char *s)
{
    char *m = strdup(s);
    if (! m) {
	fprintf(stderr, "smidump: strdup failed - running out of memory\n");
	exit(1);
    }
    return m;
}



void xfree(void *ptr)
{
    free(ptr);
}



static Module* addModule(SmiModule *smiModule, int flags)
{
    Module *mPtr;

    mPtr = xmalloc(sizeof(Module));
    mPtr->smiModule = smiModule;
    mPtr->flags = flags;
    mPtr->nextPtr = moduleList;
    moduleList = mPtr;
    return mPtr;
}



static void freeModules()
{
    Module *mPtr;

    while (moduleList) {
	mPtr = moduleList;
	moduleList = moduleList->nextPtr;
	xfree(mPtr);
    }
}



static void formats()
{
    Driver *driver = driverTable;
    
    for (driver = driverTable; driver->name; driver++) {
	fprintf(stderr, "  %-12s: %s\n", driver->name,
 		driver->descr ? driver->descr : "...");
    }
}



static void usage()
{
    fprintf(stderr,
	    "Usage: smidump [-Vhsu] [-l <level>] [-c <configfile>] [-f <format>]\n"
	    "               [-p <module>] <module_or_path>\n"
	    "-V                  show version and license information\n"
	    "-h                  show usage information\n"
	    "-s                  do not generate any comments\n"
	    "-c <configfile>     load a specific configuration file\n"
	    "-p <module>         preload <module>\n"
	    "-l <level>          set maximum level of errors and warnings\n"
	    "-f <format>         use <format> when dumping (default %s)\n"
	    "-u                  print a single united output of all modules\n"
	    "<module_or_path>    name of a MIB module or file path\n\n",
	    driverTable->name);
    fprintf(stderr, "Supported formats are:\n");
    formats();
}



static void version()
{
    printf("smidump " SMI_VERSION_STRING "\n");
}



int
main(argc, argv)
    int argc;
    char *argv[];
{
    char c;
    char *modulename;
    SmiModule *smiModule;
    int flags, smiflags, i;
    Driver *driver = driverTable;

    for (i = 1; i < argc; i++) if (strstr(argv[i], "-c") == argv[i]) break;
    if (i == argc) 
	smiInit("smidump");
    else
	smiInit(NULL);

    flags = 0;
    while ((c = getopt(argc, argv, "Vhsul:f:p:c:")) != -1) {
	switch (c) {
	case 'c':
	    smiReadConfig(optarg, "smiquery");
	    break;
	case 'V':
	    version();
	    return 0;
	case 'h':
	    usage();
	    return 0;
	case 'l':
	    smiSetErrorLevel(atoi(optarg));
	    break;
	case 's':
	    flags |= SMIDUMP_FLAG_SILENT;
	    break;
	case 'p':
	    smiLoadModule(optarg);
	    break;
	case 'u':
	    flags |= SMIDUMP_FLAG_UNITE;
	    break;
	case 'f':
	    for (driver = driverTable; driver->name; driver++) {
		if (strcasecmp(driver->name, optarg) == 0) {
		    break;
		}
	    }
	    if (!driver->name) {
		fprintf(stderr, "smidump: invalid dump format `%s'"
			" - supported formats are:\n", optarg);
		formats();
		exit(1);
	    }
	    break;
	default:
	    usage();
	    exit(1);
	}
    }

    smiflags = smiGetFlags();
    smiflags |= SMI_FLAG_ERRORS;
    smiflags |= driver->smiflags;
    smiSetFlags(smiflags);

    if (flags & SMIDUMP_FLAG_UNITE && driver->ignflags & SMIDUMP_FLAG_UNITE) {
	fprintf(stderr, "smidump: %s format does not support united output:"
		" ignoring -u\n", driver->name);
	flags = (flags & ~SMIDUMP_FLAG_UNITE);
    }

    while (optind < argc) {
	modulename = smiLoadModule(argv[optind]);
	smiModule = modulename ? smiGetModule(modulename) : NULL;
	if (smiModule) {
	    (driver->func)(addModule(smiModule, flags));
	} else {
	    fprintf(stderr, "smidump: cannot locate module `%s'\n",
		    argv[optind]);
	}
	optind++;
    }

    /*
     * Call the output driver with a NULL module name if we produce
     * united output. Output drivers supporting united output usually
     * maintain internal information and wait for this final call to
     * actually generate output (and to release allocated resources).
     */

    if (flags & SMIDUMP_FLAG_UNITE) {
	(driver->func)(addModule(NULL, flags));
    }

    freeModules();
    smiExit();
    
    return 0;
} /*  */
